src/pkcs11.c

```diff
@@ -101,11 +101,60 @@ static struct {
 #endif
 };
 
+static CK_RV module_ensure(struct pkcs11_module *pkcs, struct dbg *dbg)
+{
+	CK_C_INITIALIZE_ARGS args = {
+		.flags = CKF_OS_LOCKING_OK,
+	};
+	CK_RV ck_rv;
+	int rv;
+
+	if (!pkcs || !dbg)
+		return CKR_ARGUMENTS_BAD;
+
+	/* check state unlocked */
+	if (pkcs->state == PKCS11_INITIALIZED)
+		return CKR_OK;
+
+	rv = pthread_mutex_lock(&pkcs->mutex);
+	if (rv) {
+		ps_dbg_error(dbg, "pkcs: %p, unable to lock module: %d", pkcs, rv);
+		return CKR_CANT_LOCK;
+	}
+
+	/* check state again under lock */
+	if (pkcs->state == PKCS11_INITIALIZED) {
+		ck_rv = CKR_OK;
+		goto out;
+	}
+
+	args.pReserved = (void *)pkcs->initargs;
+	ck_rv = pkcs->fns->C_Initialize(&args);
+	if (ck_rv != CKR_OK &&
+	    ck_rv != CKR_CRYPTOKI_ALREADY_INITIALIZED) {
+		ps_dbg_error(dbg, "pkcs: %p, C_Init() failed with %d", pkcs, ck_rv);
+		goto out;
+	}
+
+	pkcs->do_finalize = (ck_rv == CKR_OK);
```

Comment:
I still wonder if we should update do_finalize in a fork-re-init case. I guess we need to discuss this a bit more....

Reply:
Just for the record: the implementation for do_finalize is correct, if we do the C_Initialize() completely lazy. If an application, which is using PKCS#11 directly and also OpenSSL with the provider, it is expected that this application is doing the C_Initialize() in its own atfork-handler for the child. In this case we're safe, because we're detecting this with our 
CKR_CRYPTOKI_ALREADY_INITIALIZED logic.

So, no changes are required here.